### DISCUSSION TITLE
Profession Versions

### DISCUSSION THREAD

**Discussion by elyukai** [elyukai/optolith-client#1100] [Ideas]
Professions are the only entries in the database that remain to have **duplicate entries**. "Duplicate" in terms of they represent the same thing but have slightly different values, e. g. often professions are introduced in Core Rules or regional sourcebooks and in rule extension publications (like Magic of Aventuria) the get *enhanced* with style special abilities or new suggested and unsuitable advantages and disadvantages. Those small changes dont really justify duplicate entries. So in order to make the database duplicate-free, I've though about a way to remove those duplicates.

Profession Versions would require a new property – I've thought about `versions` or `extensions` – which basically mirrors the profession schema with the exception of the profession identifier being replaced with a profession version identifier that can be stored. This way, you can add entries and values, even to profession variants where you can also add new ones, while maintaining one file. Since those values will be *added* to the base profession they act as a symmetric difference of two profession packages.

What do you think about that?

Open questions:

- How to determine the base profession? Is the first published version? Is it the package with the smallest number of values?

**Comment by ElidiaSeehoff** [comment]
First: the base profession should be the oldest. A "smallest common denominator" could change anytime a newer version of the profession commes around, which would incure far more monkying aroung the just using the oldest version as basis. Also we try very hard to stay in canonical order with the id's so we should propably use the same approach here.

The core idea sounds ideal, as far as i understand you want to handle it similarly to Profession variants, which seams really straight forward. There is only two potentially problematic points to me: 
As the profession version will be selectable, how is this intended to work with versions of variants?  
And, what qualifies as "duplicat"? Just the ones with the identical name or are for example "Efferdgeweihter" and "Efferdgeweihter der Siebenwindküste" also duplicates that should be unified?

**Reply by elyukai** [reply]
> First: the base profession should be the oldest. A "smallest common denominator" could change anytime a newer version of the profession commes around, which would incure far more monkying aroung the just using the oldest version as basis. Also we try very hard to stay in canonical order with the id's so we should propably use the same approach here.

This is very true.

> as far as i understand you want to handle it similarly to Profession variants, which seams really straight forward

Yeah. My current plan is that it preselects the version that makes the most use of the activated publications (so it would always preselect the version with a style SA, for example), but you could change to a different version if you prefer.

> As the profession version will be selectable, how is this intended to work with versions of variants?

Since the base schema is mirrored with additional packages, different versions of entries can override/extend existing variants. The only issue with this approach might be if different versions introduce different new variants, but it would be possible that it is ensured that different variants use different identifiers.

> what qualifies as "duplicat"? Just the ones with the identical name or are for example "Efferdgeweihter" and "Efferdgeweihter der Siebenwindküste" also duplicates that should be unified?

The former one, since the latter one might have regional changes (same as with items having different weapon advantages/disadvantages). And this makes it no longer a different version, but more of a variant instead. But variants are only the official ones so it'd need a separate entry. So, yeah, you could base this on the profession name.
